// Copyright 2023:
//   GobySoft, LLC (2013-)
//   Community contributors (see AUTHORS file)
// File authors:
//   Toby Schneider <toby@gobysoft.org>
//
//
// This file is part of the Dynamic Compact Control Language Library
// ("DCCL").
//
// DCCL is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 2.1 of the License, or
// (at your option) any later version.
//
// DCCL is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with DCCL.  If not, see <http://www.gnu.org/licenses/>.
syntax = "proto2";

import "dccl/option_extensions.proto";

package dccl.test;

message TestMsg
{
    option (dccl.msg) = {
        id: 1
        max_bytes: 32
        codec_version: 4
    };

    repeated int32 a = 1 [
        (dccl.field) = { min: -100 max: 100 max_repeat: 5 }
    ];  // default min repeat (0)
    repeated int32 b = 2 [
        (dccl.field) = { min: -100 max: 100 min_repeat: 2 max_repeat: 3 }
    ];  // min repeat and max repeat different
    repeated int32 c = 3 [
        (dccl.field) = { min: -100 max: 100 min_repeat: 3 max_repeat: 3 }
    ];  // min repeat and max repeat the same
}

message InvalidTestMsgMissingMaxRepeat
{
    option (dccl.msg) = {
        id: 2
        max_bytes: 32
        codec_version: 4
    };

    repeated int32 a = 1 [(dccl.field) = { min: -100 max: 100 }];
}

message InvalidTestMsgMaxRepeatLessThanOne
{
    option (dccl.msg) = {
        id: 2
        max_bytes: 32
        codec_version: 4
    };

    repeated int32 a = 1 [(dccl.field) = { min: -100 max: 100 max_repeat: 0 }];
}

message InvalidTestMsgMaxRepeatLessThanMinRepeat
{
    option (dccl.msg) = {
        id: 2
        max_bytes: 32
        codec_version: 4
    };

    repeated int32 a = 1
        [(dccl.field) = { min: -100 max: 100 min_repeat: 5 max_repeat: 3 }];
}
